lib/client.js

```diff
@@ -140,44 +142,70 @@ class ProtomuxRpcConnection extends SuspendResource {
     // DEVNOTE: there is no need to track timers at object level (to clear them on close):
     // closing causes the RPC clients to close, causing the request to reject
     // which triggers the finally that clears the timeout
-    const { resolve, reject, promise } = rrp()
+    const timeoutSignal = new Signal()
     const timer = setTimeout(
-      () => { reject(Errors.REQUEST_TIMEOUT()) },
+      () => {
+        timeoutSignal.notify(Errors.REQUEST_TIMEOUT())
+      },
       timeout
     )
-    promise.catch(safetyCatch) // no unhandleds
-
-    const reqProm = this._connectAndSendRequest(
-      methodName,
-      args,
-      { requestEncoding, responseEncoding, rpcTimeout: timeout } // Pass on the same timeout, so the RPC request does not stay pending forever after our timeout triggered
-    )
-    reqProm.catch(safetyCatch) // no unhandleds
+    const abortSignalPromise = timeoutSignal.wait()
+    abortSignalPromise.catch(safetyCatch)
 
     try {
-      return await Promise.race([reqProm, promise])
+      return await this._requestConcurrentLimiter.execute(async () => {
+        return await this._connectAndSendRequest(
```

Comment:
```suggestion
      return this._requestConcurrentLimiter.execute(() => {
        return this._connectAndSendRequest(
```

I believe these allocate extra promises if awaited before returning.